crates/sui-snapshot/src/writer.rs

```diff
@@ -0,0 +1,426 @@
+            manifest_file_path.clone(),
+            self.local_object_store,
+            self.remote_object_store,
+        )
+        .await?;
+        Ok(())
+    }
+    fn start_upload(
+        &self,
+        receiver: Receiver<FileMetadata>,
+        upload_concurrency: NonZeroUsize,
+    ) -> Result<JoinHandle<Result<Vec<()>, anyhow::Error>>> {
+        let remote_object_store = self.remote_object_store.clone();
+        let local_object_store = self.local_object_store.clone();
+        let local_dir_path = self.dir_path.clone();
+        let epoch_dir = self.epoch_dir();
+        let join_handle = tokio::spawn(async move {
+            let results: Vec<Result<(), anyhow::Error>> = ReceiverStream::new(receiver)
+                .map(|file_metadata| {
+                    let backoff = backoff::ExponentialBackoff::default();
+                    let file_path = file_metadata.file_path(&epoch_dir);
+                    let remote_object_store = remote_object_store.clone();
+                    let local_object_store = local_object_store.clone();
+                    let local_dir_path = local_dir_path.clone();
+                    async move {
+                        retry(backoff, || async {
+                            copy_file(
+                                file_path.clone(),
+                                file_path.clone(),
+                                local_object_store.clone(),
+                                remote_object_store.clone(),
+                            )
+                            .await
+                            .map_err(|e| anyhow!("Failed to upload file: {e}"))
+                            .map_err(backoff::Error::transient)?;
+                            fs::remove_file(path_to_filesystem(
+                                local_dir_path.clone(),
+                                &file_path.clone(),
+                            )?)
+                            .map_err(|e| anyhow!("Failed to delete file: {e}"))
+                            .map_err(backoff::Error::transient)?;
+                            Ok(())
+                        })
+                        .await?;
+                        Ok(())
+                    }
+                })
+                .boxed()
+                .buffer_unordered(upload_concurrency.get())
+                .collect()
+                .await;
+            results
+                .into_iter()
+                .collect::<Result<Vec<()>, anyhow::Error>>()
+        });
+        Ok(join_handle)
+    }
+    fn write_live_object_set<F>(
+        &mut self,
+        perpetual_db: Arc<AuthorityPerpetualTables>,
+        sender: Sender<FileMetadata>,
+        bucket_func: F,
+    ) -> Result<Vec<FileMetadata>>
+    where
+        F: Fn(&LiveObject) -> u32,
+    {
+        let mut object_writers: HashMap<u32, LiveObjectSetWriterV1> = HashMap::new();
+        for object in perpetual_db.iter_live_object_set() {
+            let bucket_num = bucket_func(&object);
```

Comment:
Since we're already paying the cost here of iterating through the entire live object set, wondering if it would make sense to accumulate these objects (we already [have this function](https://github.com/MystenLabs/sui/blob/09b2081498366df936abae26eea4b2d5cafb2788/crates/sui-core/src/state_accumulator.rs#L367) that does all the work) and return the accumulator as a commitment to what was written. Then a callsite can easily have an invariant to check against the end of epoch root state hash. That way we discover issues with the snapshot at creation time rather than consumption time

Reply:
Yeah i think it's not a bad idea to add that invariant here by passing in the root hash. Definitely feel free to add it!